[022] add internships relations

Maksim-maevsky opened the pull request:
- added enums for format and skill 
- added tables user-skill and internship skill
- added initialization scripts for  user-skill and internship skill

Maksim-maevsky committed:
added format, technology, internship-fk, like liquibase, and initialization for them.

Maksim-maevsky committed:
added technology entity.

Maksim-maevsky committed:
returned variables for application.yaml

Maksim-maevsky committed:
Merge branch 'development' into 022-add-internships-relations

# Conflicts:
#	src/main/resources/application.yaml

mortal-gpt commented on src/main/java/by/exadel/internship/entity/Internship.java:
```diff
@@ -55,11 +53,17 @@
-//    private List<Technology> technologyList;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "inship_format_name", nullable = false)
+    private InternshipFormat InternshipFormat;
```
```suggestion
    private InternshipFormat internshipFormat;
```

Maksim-maevsky replied:
done

mortal-gpt commented on src/main/java/by/exadel/internship/entity/Internship.java:
```diff
@@ -55,11 +53,17 @@
+
+    @ManyToOne
+    @JoinColumn(name="inship_technology_name", nullable=false)
+    private Technology technology;
```
Just note: Did you investigate posibility to use  @ElementCollection?

Maksim-maevsky replied:
done 
![image](https://user-images.githubusercontent.com/65280608/114389871-3918d980-9b9e-11eb-9c81-253225a7e1e1.png)

Maksim-maevsky replied:
![image](https://user-images.githubusercontent.com/65280608/114399517-ac741880-9ba9-11eb-8861-70acff354406.png)

mortal-gpt commented on src/main/resources/application.yaml:
```diff
@@ -2,20 +2,23 @@ spring:
     show-sql: true
     hibernate:
-      ddl-auto: validate
+      ddl-auto: update
```
revert configuration changes please

Maksim-maevsky replied:
done

kzaretskaya commented on src/main/java/by/exadel/internship/entity/Internship.java:
```diff
@@ -55,11 +53,17 @@
-//    private InternshipFormat InternshipFormat;
-//    private List<Technology> technologyList;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "inship_format_name", nullable = false)
```
Why `nullable = false`? I think when admin create internship it's possible that format can not be approved yet.

Maksim-maevsky replied:
I am going to text  Zanna today about it with other questions.

Maksim-maevsky replied:
done like nullable to Internship and  db

Maksim-maevsky committed:
removed Format table

Maksim-maevsky committed:
added schema for @Table User.
rename variable InternshipFormat
added Type Mapper for Enum for post in DB

kzaretskaya commented on src/main/resources/application.yaml:
```diff
@@ -2,16 +2,17 @@ spring:
-    url: ${SPRING_DATASOURCE_URL}
-    password: ${SPRING_DATASOURSE_PASSWORD}
-    username: ${SPRING_DATASOURSE_USERNAME}
+    url: jdbc:postgresql://localhost:5432/postgres
```
revert to previous state

Maksim-maevsky replied:
done

kzaretskaya commented on src/main/java/by/exadel/internship/entity/user/User.java:
```diff
@@ -1,20 +1,20 @@
 
 @Entity
-@Table(name = "user")
+@Table(name = "user", schema = "public")
```
what purposes do you specify schema for?

Maksim-maevsky replied:
IT WAS Grammar exception from hibrnate, and it was pretty strange, because everything was right, and schema was decision from net about this problem.

Fenydude replied:
I think that because postgres has another user table and schema = "public" resolves that conflict

Maksim-maevsky replied:
yes, but we took desicion, how i remember, about singular naming. @kzaretskaya  maybe we should change this?

kzaretskaya replied:
What will be changed? to plural naming? no. This entity is already in branch development, how does it work there? If problem appeared on your branch it need to figure out cause.

Maksim-maevsky replied:
![image](https://user-images.githubusercontent.com/65280608/114516336-f9a5c800-9c45-11eb-9b74-bf8fb9d57457.png)
this is a working option also

Maksim-maevsky replied:
https://stackoverflow.com/questions/22256124/cannot-create-a-database-table-named-user-in-postgresql

user is a reserved word and it's usually not a good idea use reserved words for identifiers (tables, columns).

kzaretskaya replied:
I think it's not a good idea to wrap with quotes. This problem had to be appeared when the entity was created (probably it's not tested).
My suggestion to rename table to user_detail.
@viktarkolbik @SergeyKoval what do you think?

SergeyKoval replied:
Agree with @kzaretskaya

Maksim-maevsky replied:
done 
![image](https://user-images.githubusercontent.com/65280608/114573224-dba98900-9c80-11eb-9004-4ce851fc1c11.png)

Maksim-maevsky committed:
renamed technology

Maksim-maevsky committed:
reverted url pass username variable in application.yaml

Maksim-maevsky committed:
reverted validate application.yaml

Maksim-maevsky committed:
created v.0.3  for liquibase, master, changelog-skill, imports with initialization scripts for internship-skill table, enum Skill.

added @@ElementCollection for Internship

Maksim-maevsky committed:
create Location entity
added user-skill connection and initialization
also used @ElementCollection

Maksim-maevsky committed:
added @CollectionTable for entity user and internship

mortal-gpt commented on src/main/java/by/exadel/internship/entity/User.java:
```diff
@@ -38,10 +40,13 @@
+    @Column(name="skill_name", nullable=false)
+    @Enumerated(EnumType.STRING)
+    @ElementCollection(fetch = FetchType.EAGER)
+    @CollectionTable(name = "user_skill", joinColumns = @JoinColumn(name = "u_s_u_id"))
```
```suggestion
    @CollectionTable(name = "user_skill", joinColumns = @JoinColumn(name = "us_u_id"))
```

Maksim-maevsky replied:
done

mortal-gpt commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+                  name: skill_name
+                  type: SKILL
+                  constraints:
+                    nullable: false
```
```suggestion
                    nullable: true
```

skill_name can't exist without is_inship_id

Maksim-maevsky replied:
done

Maksim-maevsky replied:
maybe false is right?

SergeyKoval replied:
Looks like false is correct

mortal-gpt commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+                  constraints:
+                    nullable: false
+              - column:
+                  name: i_s_inship_id
```
name: is_inship_id

Maksim-maevsky replied:
done

mortal-gpt commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+            tableName: user_skill
+            columns:
+              - column:
+                  name: u_s_u_id
```
```suggestion
                  name: us_u_id
```

Maksim-maevsky replied:
done

mortal-gpt commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+                  constraints:
+                    nullable: true
+              - column:
+                  name: skill_name
```
```suggestion
                  name: us_name
```

Maksim-maevsky replied:
done

mortal-gpt commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+                  name: u_s_u_id
+                  type: UUID
+                  constraints:
+                    nullable: true
```
```suggestion
                    nullable: false
```

mortal-gpt replied:
us_u_id doesn't exist without skill_name

Maksim-maevsky replied:
done

Maksim-maevsky committed:
rename columns for user-skill and internship-skill tables

kzaretskaya commented on src/main/resources/db/changelog/v.0.3/db.changelog-skill.yaml:
```diff
@@ -0,0 +1,64 @@
+      author: maksim_maevsky
+      changes:
+        - createTable:
+            tableName: user_skill
```
Please, add to both table uniqie constraint to avoid set the same skill to one user/internship.

Maksim-maevsky replied:
done
![image](https://user-images.githubusercontent.com/65280608/114505098-8ba6d400-9c38-11eb-9d7d-205cdf83c46b.png)

kzaretskaya commented on src/main/java/by/exadel/internship/entity/Internship.java:
```diff
@@ -55,11 +55,20 @@
-//    private List<Technology> technologyList;
+    @Column(name = "inship_format_name", nullable = false)
+    @Enumerated(EnumType.STRING)
+    @Type(type = "by.exadel.internship.mapper.enum_mapper.EnumTypePostgreSQL")
```
why do you specify type here?

Maksim-maevsky replied:
it is for the future , when we will add some  Inrernship.

Maksim-maevsky replied:
we will think about how we could use it from one place, thanks

kzaretskaya commented on src/main/java/by/exadel/internship/entity/Location.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class Location {
```
Remove please the class from this PR. It'll be added in separate PR and with all appropriate annotation.

Maksim-maevsky replied:
done

Maksim-maevsky committed:
removed Location entity

Maksim-maevsky committed:
added uniqueConstraint for user-skill and internship_skill

Maksim-maevsky committed:
removed schema to user entity

SergeyKoval commented on src/main/java/by/exadel/internship/entity/User.java:
```diff
@@ -38,10 +40,13 @@
-//    private List<Technology> techTechnology;
+    @Column(name="us_name", nullable=false)
+    @Enumerated(EnumType.STRING)
+    @ElementCollection(fetch = FetchType.EAGER)
```
Do you really need EAGER here? 
I suggest use LAZY by default as much as possible. 

Here I expect a lot of scenarios where this list of skills will be redundant. For example to assign user for the interview you'll need to get all users with required skill. Getting all skills for each user will be redundant.

Maksim-maevsky replied:
i got it. Done.

kzaretskaya commented on src/main/java/by/exadel/internship/entity/User.java:
```diff
@@ -1,20 +1,22 @@
 
 @Entity
-@Table(name = "user")
+@Table(name="\"user\"")
```
I think it's not a good idea to wrap with quotes. This problem had to be appeared when the entity was created (probably it's not tested).
My suggestion to rename table to `user_detail`.
@viktarkolbik @SergeyKoval what do you think?

Maksim-maevsky committed:
renamed table user
added inship_format_name like nullable
added LAZY fetch instead EAGER for user and internship

Maksim-maevsky committed:
Merge branch 'development' into 022-add-internships-relations

# Conflicts:
#	src/main/java/by/exadel/internship/entity/Internship.java
#	src/main/java/by/exadel/internship/service/InternshipService.java
#	src/main/resources/db/changelog/db.changelog-master.yaml
#	src/main/resources/db/changelog/v.0.3/db.changelog-master.yaml

Maksim-maevsky committed:
resolved problems with development

SergeyKoval reviewed (approved):
Don't forget to reduce number of sql queriws
